core/domain/prod_validation_jobs_one_off.py

```diff
@@ -2474,213 +2600,1326 @@ def _get_custom_validation_functions(cls):
+            cls._validate_user_id_not_in_subscriber_ids,
+            cls._validate_user_id_in_creator_ids]
+
+
+class UserRecentChangesBatchModelValidator(BaseUserModelValidator):
+    """Class for validating UserRecentChangesBatchModels."""
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {
+            'user_settings_ids': (
+                user_models.UserSettingsModel, [item.id])
+        }
+
+    @classmethod
+    def _validate_job_queued_msec(cls, item):
+        """Validates that job queued msec is less than the time
+        when job was run.
+
+        Args:
+            item: ndb.Model. UserRecentChangesBatchModel to validate.
+        """
+        current_msec = utils.get_current_time_in_millisecs()
+        if item.job_queued_msec > current_msec:
+            cls.errors['job queued msec check'].append(
+                'Entity id %s: job queued msec %s is greater than '
+                'the time when job was run' % (
+                    item.id, item.job_queued_msec))
+
+    @classmethod
+    def _get_custom_validation_functions(cls):
+        return [cls._validate_job_queued_msec]
+
+
+class UserStatsModelValidator(BaseUserModelValidator):
+    """Class for validating UserStatsModels."""
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {
+            'user_settings_ids': (
+                user_models.UserSettingsModel, [item.id])
+        }
+
+    @classmethod
+    def _validate_schema_version(cls, item):
+        """Validates that schema version is less than current version.
+
+        Args:
+            item: ndb.Model. UserStatsModel to validate.
+        """
+        if item.schema_version > feconf.CURRENT_DASHBOARD_STATS_SCHEMA_VERSION:
+            cls.errors['schema version check'].append(
+                'Entity id %s: schema version %s is greater than '
+                'current version %s' % (
+                    item.id, item.schema_version,
+                    feconf.CURRENT_DASHBOARD_STATS_SCHEMA_VERSION))
+
+    @classmethod
+    def _validate_weekly_creator_stats_list(cls, item):
+        """Validates that each item weekly_creator_stats_list is keyed
+        by a datetime field and value as a dict with keys: num_ratings,
+        average_ratings, total_plays. Values for these keys should be
+        integers.
+
+        Args:
+            item: ndb.Model. UserStatsModel to validate.
+        """
+        current_time_str = datetime.datetime.utcnow().strftime(
+            feconf.DASHBOARD_STATS_DATETIME_STRING_FORMAT)
+        for stat in item.weekly_creator_stats_list:
+            for key, value in stat.iteritems():
+                is_invalid = False
+                try:
+                    datetime.datetime.strptime(
+                        key, feconf.DASHBOARD_STATS_DATETIME_STRING_FORMAT)
+                except Exception:
+                    is_invalid = True
+                if not is_invalid and key > current_time_str:
```

Comment:
I think, instead of carrying is_invalid around, you could just use assertions within the try/except. That might lead to cleaner code.

Reply:
Done.